Interface tester/tester plugin (#57)
canonical/charm-relation-interfaces: README_INTERFACE_TESTS.md

# Interface tests

Charm relation interfaces are information exchange protocols for Juju charms.

> In this repository, under each interface specification, one can find a list of interface tests. Their goal is to provide an as strict as possible set of rules to determine whether a given charm satisfies that interface protocol.

Reviewer: "In this repository, under each interface specification, one can find a list of interface tests." is not exactly True

> # Interface tests

Reviewer: This page is pretty hard to follow. It serves as a tutorial, reference, explanation, how-to but without clear delineations. 

You may want to make those separations clearer with subections called "tutorial (or getting started)" and "explanation" or simply get rid of the exmplanations and focus on the tutorial.

Author: I'll get started on cleaning it up, meanwhile could you check that the remaining issues are resolved?

Author: there should be a decent howto now.

Charm relation interfaces are information exchange protocols for Juju charms.
In this repository, under each interface specification, it is possible to store a list of interface tests. The interface maintainers will write and maintain those tests for the interfaces they own. 

> The goal of the tests is to provide an as strict as possible set of rules to determine whether a given charm satisfies that interface protocol.

Reviewer: ```suggestion
The goal of the interface tests is to determine whether a given charm satisfies an interface protocol.
```

ingress = Relation(
    'ingress',
    remote_app_data={"data": "foo"}  # whatever data the provider expects to be supplied
)
state_out = State(relations=[ingress]).trigger(ingress.joined_event, ...)
assert state_out.get_relations('ingress')[0].local_app_data  # not empty

# and: 
ingress = Relation(
    'ingress', # the requirer has not sent its side of the data!
)
state_out = State(relations=[ingress]).trigger(ingress.joined_event, ...)
assert not state_out.get_relations('ingress')[0].local_app_data  # empty! 
```

Therefore, each relation interface specification translates to two sets of tests: some are meant for the requirer, some for the provider. 

In order to add interface tests to a charm relation interface `"my-interface"`, you should commit to this repository one or more python files under `repo-root/interfaces/my-interface/v0/interface_tests/`; for example:
`repo-root/interfaces/my-interface/v0/interface_tests/requirer_tests.py` and `repo-root/interfaces/my-interface/v0/interface_tests/provider_tests.py`.

Any python file in that directory will be scraped for tests. A test is any function decorated with the `interface_tester.interface_test_case` decorator.

> Before you start writing interface tests, it's a good idea to pip install [interface-tester-pytest](https://github.com/PietroPasotti/interface-tester-pytest).

Reviewer: This should point to the canonical org's project

- `interface_tester.interface_test.SchemaConfig.skip` (or the string`"skip"`): skip schema validation for this test; use sparingly.
- `interface_tester.interface_test.SchemaConfig.empty` (or the string`"empty"`): assert that any `ingress` relation found in the `state_out` has **no relation data** at all (local side).
- you can pass a custom `interface_tester.schema_base.DataBagSchema` subclass, which will be used to validate any `ingress` relation found in `state_out`. This will replace the default one found in `schema.py` for this test only.

## Steps of an interface test
An interface test such as the one above has three steps:

- verify that the scenario runs
- verify that the output state (`state_out`) is valid (by the test-writer's definition)
- validate against the relation schemas

Failure in any of the three steps means a failed test.

# Matrix-testing interface compliance
Suppose we have an interface `foo`, with:
- `charms.yaml` listing 2 providers and 2 requirers of the `foo` interface.
- `schema.py` detailing the interface schema (optional: schema validation will be skipped if not found)
- `interface_tests/*.py` providing a list of interface tests for both roles

> You can then run `python ./run_matrix.py foo`.

Reviewer: It looks like run_matrix doesn't accept arguments as it is

This will attempt to run the interface tests on all charms in `foo/charms.yaml`.
Omitting the `foo` argument will run the tests for all interfaces (warning: might take some time.)

# Using interface tests as a charm author

> If you are the owner of a charm, and you are interested in adding your charm to an interface's `charms.yaml` (and thereby 'prove' that your charm complies with a given relation interface specification), then you need to follow the steps detailed at [interface-tester-pytest](https://github.com/PietroPasotti/interface-tester-pytest).

Reviewer: This should point to the canonical org's project

# Interface tests

Charm relation interfaces are information exchange protocols for Juju charms.
In this repository, under each interface specification, it is possible to store a list of interface tests. The interface maintainers will write and maintain those tests for the interfaces they own. 
The goal of the interface tests is to determine whether a given charm satisfies an interface protocol.

There are two personas that need to work with interface tests: interface authors and interface users.
If you are an interface user, i.e. the owner of a charm that uses some of the interfaces specified in this repository, then you should look at the documentation in the [interface-tester-pytest](https://github.com/canonical/interface-tester-pytest) repository.

If you are an interface author, this page is for you.

# How-To write interface tests as an interface author

> Firstly, `pip install` [`interface-tester-pytest`](https://github.com/canonical/interface-tester-pytest).

Reviewer: Firstly, install the interface tester package:
```
pip install git+https://github.com/canonical/interface-tester-pytest
```

## A minimal test case

Specify the schema for the `ingress` interface databags in `./interfaces/ingress/v0/schema.py`.

Create a python file at `<repo-root>/interfaces/ingress/v0/interface_tests/my_tests.py` with these contents:

```python
from scenario import State
from interface_tester.interface_test import interface_test_case

@interface_test_case(
    event='ingress-relation-joined',
    role='requirer',
)
def test_data_published_on_joined(output_state: State):
    return
```

This is a complete interface test.

## Validating the output state

If the specification dictates something more than "the databag contents should match this schema in that event", for example:

> on `ingress-relation-joined`, the requirer should write its part of the data to the relation and set its status to Active.

The test would then become:

```python
from scenario import State
from interface_tester.interface_test import interface_test_case

@interface_test_case(
    event='ingress-relation-joined',
    role='requirer',
)
def test_data_published_on_joined(output_state: State):
    assert output_state.status.unit.name == 'active'
```

> 

Reviewer: We should provide the command to actually run the test

## Testing a negative

The second norm we have above contains a negative. We split it into two assertions to write the tests:

> on `ingress-relation-changed`, if the requirer has provided 'data' and 'baz' information in its relation databags, then the provider should publish its part of the data.
> on `ingress-relation-changed`, if the requirer has NOT provided 'data' and 'baz' information in its relation databags, then the provider should publish no data.

This becomes two test cases:

```python
from scenario import State, Relation
from interface_tester.interface_test import interface_test_case, SchemaConfig

@interface_test_case(
    event='ingress-relation-joined',
    role='provider',
    input_state=State(
        relations=[Relation(
            endpoint='foo',
            interface='ingress',
            remote_app_name='remote',  # this is our simulated requirer
            remote_app_data={
                'data': 'foo-bar',
                'baz': 'qux'
            }
        )]
    )
)
def test_data_published_on_joined_if_remote_has_sent_valid_data(output_state: State):

>     pass

Reviewer: I know this is out of scope of this PR but the framework makes it non-trivial to check what is exactly validated. We're not asserting anything explicitly.

When I'm reviewing this test for example, I have a hard time figuring out, what exactly are we validating. The same goes for the next one where the devil's in the "schema=SchemaConfig.empty" detail.

Author: That's a good point, we can revisit the design of the tests later

Author: opened a new issue to track this

    
@interface_test_case(
    event='ingress-relation-joined',
    role='provider',
    schema=SchemaConfig.empty,
    input_state=State(
        relations=[Relation(
            endpoint='foo',
            interface='ingress',
            remote_app_name='remote',
            remote_app_data={
                'some': 'rubbish'
            }
        )]
    )

> 

Reviewer: ```suggestion
```

Reviewer: This comment was resolved but the fix wasn't applied

(in general, please resolve comments after the fix was made or both parties agreed on it not being fixed)

)
def test_no_data_published_on_joined_if_remote_has_not_sent_valid_data(output_state: State):
    pass
```

The first test verifies that, if the requirer has provided correct data, then the provider will populate its side of the databag.
The second test verifies the negative side of the norm.
Note the usage of `SchemaConfig.empty`. That is what disables the 'default' schema validation and instructs the test runner to verify that the provider-side databags are empty instead of containing whatever they should contain according to `schema.py`.  

## Verifying the test is discoverable

> If you run `interface_tester discover` in a shell (in the repo root; or check out `--help` to see how to customize that), you should see something like a list of:

Reviewer: We're still in the `how-to` section here so I'd continue with " run X", not "if you ..."

> If you run `interface_tester discover` in a shell (in the repo root; or check out `--help` to see how to customize that), you should see something like a list of:

Reviewer: We're still in the `how-to` section here so I'd continue with " run X", not "if you ..."

```yaml

> my-interface:

Reviewer: This should match with the `ingress` example

  - v0:
      - requirer:
          - tests:
            - <test name>:: <event name>
          - schema: OK
          - charms:
              - <charm name> (<url>)
```

> If you are adding tests to an interface, you should be able to see them listed under `tests`. 

Reviewer: Are those 3 sentences needd in the "how-to"?

If your interface has a `schema.py`, you can check whether the schema is specified correctly or not.
If your interface lists one or more charms in `charms.yaml`, you can verify here that the format is correct.

# Reference: how does it work?
Each interface test maps to a [Scenario test](https://github.com/canonical/ops-scenario).

The metadata passed to `interface_test_case`, along with metadata gathered from the charm being tested, is used to assemble a scenario test. Once that is done, each interface test can be broken down in three steps, each one verifying separate things in order:

- verify that the scenario test runs (i.e. it can produce an output state without the charm raising exceptions)
- verify that the output state is valid (by the interface-test-writer's definition)
- validate the local relation databags against the role's relation schema provided in `schema.py`

If any of these steps fail, the test as a whole is considered failed.

## Configuring the schema to be used in an interface test

> This means that the test you just added is being collected. `state` indicates the presence of a custom `input_state` (which we did pass in the example above). `schema` indicates the type of schema validation to be performed.

Reviewer: It's strange to start the block with "This means..."

Any python file in that directory will be scraped for tests. A test is any function decorated with the `interface_tester.interface_test_case` decorator.
Before you start writing interface tests, it's a good idea to pip install [interface-tester-pytest](https://github.com/canonical/interface-tester-pytest).
That will give you the right type hints as you're typing the code, and will also give you a little utility cli to verify that the tests you are writing are being correctly picked up by the same code that will eventually be used to run them.

Suppose you want to contribute interface tests for the `ingress` interface.
Some tests are meant for the requirer, some for the provider. 

Simplifying a bit, the `ingress` protocol implemented by e.g. Traefik contains some norms for the requirer, and some for the provider:
- on `ingress-relation-joined`, the requirer should write its part of the data to the relation. 
- on `ingress-relation-changed`, if the provider has written its part of the data to the relation, then the provider should publish its part of the data; otherwise it should publish no data.

## A minimal test case

> An interface test case is the direct translation of one such statement into a scenario test. For example:

Reviewer: Avoid explanations in a how-to. The goal of a how-to is not to learn stuff but to help in development.

The goal of the interface tests is to determine whether a given charm satisfies an interface protocol.

There are two personas that need to work with interface tests: interface authors and interface users.
If you are an interface user, i.e. the owner of a charm that uses some of the interfaces specified in this repository, then you should look at the documentation in the [interface-tester-pytest](https://github.com/canonical/interface-tester-pytest) repository.

If you are an interface author, this page is for you.

# How-To write interface tests as an interface author

Throughout this example we will write some interface tests for the `ingress` interface.
This is the expected behaviour for the interface:
1) on `ingress-relation-joined`, the requirer should write its part of the data to the relation. 
2) on `ingress-relation-changed`, if the provider has written its part of the data to the relation, then the provider should publish its part of the data; otherwise it should publish no data.

## A minimal test case

> Install the [interface tester package](https://github.com/canonical/interface-tester-pytest).:

Reviewer: ```suggestion
Install the [interface tester package](https://github.com/canonical/interface-tester-pytest):
```

`pip install git+https://github.com/canonical/interface-tester-pytest`

> Specify the schema for the `ingress` interface databags in `./interfaces/ingress/v0/schema.py`.

Author: Done, I've added one. Let me know if it's good enough, I was planning to have separate documentation for that.

> Specify the schema for the `ingress` interface databags in `./interfaces/ingress/v0/schema.py`.

Author: As in: we can expand elsewhere and link from here, but it's good if this howto is self-contained.

> Create a python file at `<repo-root>/interfaces/ingress/v0/interface_tests/my_tests.py` with these contents:

Reviewer: ```suggestion
Create a python file at `<repo-root>/interfaces/ingress/v0/interface_tests/my_tests.py` with this content:
```

```python
from scenario import State
from interface_tester.interface_test import interface_test_case

@interface_test_case(
    event='ingress-relation-joined',
    role='requirer',
)
def test_data_published_on_joined(output_state: State):
    return
```

> Run the test with `python run_matrix.py --include ingress`.

Reviewer: Suggestion (sorry it's impossible to add suggestions with ```'s in them):

Run the test:
```
python run_matrix.py --include ingress
```

There are two personas that need to work with interface tests: interface authors and interface users.
If you are an interface user, i.e. the owner of a charm that uses some of the interfaces specified in this repository, then you should look at the documentation in the [interface-tester-pytest](https://github.com/canonical/interface-tester-pytest) repository.

If you are an interface author, this page is for you.

# How-To write interface tests as an interface author

Throughout this example we will write some interface tests for the `ingress` interface.
This is the expected behaviour for the interface:
1) on `ingress-relation-joined`, the requirer should write its part of the data to the relation. 
2) on `ingress-relation-changed`, if the provider has written its part of the data to the relation, then the provider should publish its part of the data; otherwise it should publish no data.

## A minimal test case

### Register a charm to the interface
Write to `./interfaces/ingress/v0/charms.yaml`:

```yaml
providers:
  - name: traefik-k8s
    url: https://github.com/canonical/traefik-k8s-operator

requirers: []
```

Verify that the `charms.yaml` file format is correct:

> > interface_tester discover --include ingress

Reviewer: Use code blocks instead of blockquotes for terminal commands. This is applicable at a couple of places

Author: Fixed throughout
